File: lib/school/nenep.ex
defmodule School.NeNep do
  use Ecto.Schema
  alias School.NeNep
  import Ecto.Changeset

  schema "nenep" do
    field :day, :integer
    field :week_start_date, :integer
    belongs_to :class, School.Class, foreign_key: :class_id

    field :siso, :integer
    field :khanquang, :integer
    field :dongphuc, :integer
    field :vesinh, :integer

    field :chaoco, :integer
    field :truybai, :integer
    field :bakhong, :integer
    field :shdoi, :integer

    field :tdvuichoi, :integer
    field :nghithucdoi, :integer
    field :ghichu, :string
  end

  defp validate_privilege(hoctap, map, account) do
    check_privilege = fn 
      changeset, true, _, _, _ -> changeset
      changeset, _, true, date, date -> changeset
      changeset, _, _, _, _ -> 
        add_error(changeset, :privilege, "Does not have the privilege")
    end

    hoctap 
    |> cast(map, [:class_id])
    |> validate_required([:class_id])
    |> check_privilege.(
      account.is_admin, account.is_saodo,
      map[:current_week], map[:week_start_date]
    )
  end

  defp validate_values(hoctap, map) do
    hoctap
    |> cast(map, [
      :siso, :khanquang, :dongphuc, :vesinh, 
      :chaoco, :truybai, :bakhong, :shdoi, 
      :tdvuichoi, :nghithucdoi, :ghichu
    ], empty_values: [])
    |> validate_required([
      :siso, :khanquang, :dongphuc, :vesinh, 
      :chaoco, :truybai, :bakhong, :shdoi, 
      :tdvuichoi, :nghithucdoi
    ])
    |> validate_length(:ghichu, max: 100)

    |> validate_number(:siso, greater_than_or_equal_to: 0)
    |> validate_number(:khanquang, greater_than_or_equal_to: 0)
    |> validate_number(:dongphuc, greater_than_or_equal_to: 0)
    |> validate_number(:vesinh, greater_than_or_equal_to: 0)

    |> validate_number(:chaoco, greater_than_or_equal_to: 0)
    |> validate_number(:truybai, greater_than_or_equal_to: 0)
    |> validate_number(:bakhong, greater_than_or_equal_to: 0)
    |> validate_number(:shdoi, greater_than_or_equal_to: 0)

    |> validate_number(:tdvuichoi, greater_than_or_equal_to: 0)
    |> validate_number(:nghithucdoi, greater_than_or_equal_to: 0)
  end

  def new(map, account) do 
    %NeNep{}
    |> validate_privilege(map, account)
    |> School.HocTap.validate_date(map)
    |> validate_values(map)
  end

  def update(map, account) do 
    %NeNep{id: map.id}
    |> validate_privilege(map, account)
    |> School.HocTap.validate_date(map)
    |> validate_values(map)
  end

  def delete(map, account) do
    %NeNep{id: map.id}
    |> validate_privilege(map, account)
  end
end

File: assets/static/js/nenep.js
function get_value(e, name) {
    input = e.find("input[name=" + name + "]");
    return input.val();
}

function get_map(e) {
    map = {
        "_csrf_token": get_value(e, "_csrf_token"),
        "id": get_value(e, "id"),
        "class_id": get_value(e, "class_id"),

        "day": get_value(e, "day"),
        "week_start_date": get_value(e, "week_start_date"),

        "siso": get_value(e, "siso"),
        "khanquang": get_value(e, "khanquang"),
        "dongphuc": get_value(e, "dongphuc"),
        "vesinh": get_value(e, "vesinh"),

        "chaoco": get_value(e, "chaoco"),
        "truybai": get_value(e, "truybai"),
        "bakhong": get_value(e, "bakhong"),
        "shdoi": get_value(e, "shdoi"),

        "tdvuichoi": get_value(e, "tdvuichoi"),
        "nghithucdoi": get_value(e, "nghithucdoi"),
        "ghichu": get_value(e, "ghichu"),
    };
    return map;
}

function update_button(e) {
    $.post("/nenep/update/", get_map(e))
        .always(function() { location.reload(); });
}

function delete_button(e) {
    $.post("/nenep/delete/", get_map(e))
        .always(function() { location.reload(); });
}

function add_button(e) {
    $.post("/nenep/add/", get_map(e))
        .always(function() { location.reload(); });
}
